In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data = "https://archive.ics.uci.edu/static/public/222/bank+marketing.zip"

In [8]:
!wget $data

--2024-10-07 23:01:51--  https://archive.ics.uci.edu/static/public/222/bank+marketing.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘bank+marketing.zip’

bank+marketing.zip      [       <=>          ] 999.85K   615KB/s    in 1.6s    

2024-10-07 23:01:54 (615 KB/s) - ‘bank+marketing.zip’ saved [1023843]



In [11]:
!unzip bank+marketing.zip

Archive:  bank+marketing.zip
 extracting: bank.zip                
 extracting: bank-additional.zip     


In [12]:
!unzip bank.zip -d bank
!unzip bank-additional.zip -d bank-additional

Archive:  bank.zip
  inflating: bank/bank-full.csv      
  inflating: bank/bank-names.txt     
  inflating: bank/bank.csv           
Archive:  bank-additional.zip
   creating: bank-additional/bank-additional/
  inflating: bank-additional/bank-additional/.DS_Store  
   creating: bank-additional/__MACOSX/
   creating: bank-additional/__MACOSX/bank-additional/
  inflating: bank-additional/__MACOSX/bank-additional/._.DS_Store  
  inflating: bank-additional/bank-additional/.Rhistory  
  inflating: bank-additional/bank-additional/bank-additional-full.csv  
  inflating: bank-additional/bank-additional/bank-additional-names.txt  
  inflating: bank-additional/bank-additional/bank-additional.csv  
  inflating: bank-additional/__MACOSX/._bank-additional  


In [3]:
df = pd.read_csv('./bank/bank-full.csv', delimiter=';')
df.shape

(45211, 17)

In [4]:
df.head()

age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married    unknown      no     1506     yes   no   
4   33       unknown   single    unknown      no        1      no   no   

   contact  day month  duration  campaign  pdays  previous poutcome   y  
0  unknown    5   may       261         1     -1         0  unknown  no  
1  unknown    5   may       151         1     -1         0  unknown  no  
2  unknown    5   may        76         1     -1         0  unknown  no  
3  unknown    5   may        92         1     -1         0  unknown  no  
4  unknown    5   may       198         1     -1         0  unknown  no

In [5]:
df.T.head()

0           1             2            3        4      \
age                58          44            33           47       33   
job        management  technician  entrepreneur  blue-collar  unknown   
marital       married      single       married      married   single   
education    tertiary   secondary     secondary      unknown  unknown   
default            no          no            no           no       no   

                5           6             7        8           9      ...  \
age                35          28            42       58          43  ...   
job        management  management  entrepreneur  retired  technician  ...   
marital       married      single      divorced  married      single  ...   
education    tertiary    tertiary      tertiary  primary   secondary  ...   
default            no          no           yes       no          no  ...   

                45201      45202     45203      45204       45205       45206  \
age                53         34        23         73          25          51   
job        management     admin.   student    retired  technician  technician   
marital       married     single    single    married      single     married   
education    tertiary  secondary  tertiary  secondary   secondary    tertiary   
default            no         no        no         no          no          no   

              45207      45208        45209         45210  
age              71         72           57            37  
job         retired    retired  blue-collar  entrepreneur  
marital    divorced    married      married       married  
education   primary  secondary    secondary     secondary  
default          no         no           no            no  

[5 rows x 45211 columns]

In [6]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'y'],
      dtype='object')

In [7]:
useful_columns =  ['age', 'job', 'marital', 'education', 'balance', 'housing', 'contact', 'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'y']

In [8]:
df = df[useful_columns]
df.shape

(45211, 15)

In [9]:
df.isnull().sum()

age          0
job          0
marital      0
education    0
balance      0
housing      0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [10]:
# no missing value

In [11]:
df.describe()

age        balance           day      duration      campaign  \
count  45211.000000   45211.000000  45211.000000  45211.000000  45211.000000   
mean      40.936210    1362.272058     15.806419    258.163080      2.763841   
std       10.618762    3044.765829      8.322476    257.527812      3.098021   
min       18.000000   -8019.000000      1.000000      0.000000      1.000000   
25%       33.000000      72.000000      8.000000    103.000000      1.000000   
50%       39.000000     448.000000     16.000000    180.000000      2.000000   
75%       48.000000    1428.000000     21.000000    319.000000      3.000000   
max       95.000000  102127.000000     31.000000   4918.000000     63.000000   

              pdays      previous  
count  45211.000000  45211.000000  
mean      40.197828      0.580323  
std      100.128746      2.303441  
min       -1.000000      0.000000  
25%       -1.000000      0.000000  
50%       -1.000000      0.000000  
75%       -1.000000      0.000000  
max      871.000000    275.000000

In [12]:
df.columns = df.columns.str.lower().str.replace(' ', '_')
categorial_columns = list(df.dtypes[df.dtypes == 'object'].index)
for c in categorial_columns:
    df[c] = df[c].str.lower().replace(' ', '_')
categorial_columns

['job', 'marital', 'education', 'housing', 'contact', 'month', 'poutcome', 'y']

In [13]:
df.education.value_counts()

education
secondary    23202
tertiary     13301
primary       6851
unknown       1857
Name: count, dtype: int64

In [14]:
# 1. Secondary is the mode for edcuation column

In [15]:
df.y.value_counts()

y
no     39922
yes     5289
Name: count, dtype: int64

In [16]:
df.dtypes

age           int64
job          object
marital      object
education    object
balance       int64
housing      object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y            object
dtype: object

In [17]:
numerical = df.dtypes[df.dtypes == 'int64'].index
numerical

Index(['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous'], dtype='object')

In [18]:
categorical = df.dtypes[df.dtypes == 'object'].index
categorical

Index(['job', 'marital', 'education', 'housing', 'contact', 'month',
       'poutcome', 'y'],
      dtype='object')

In [19]:
# 2. Correcation matrix amoung numerical featues
df[numerical].corr()

age   balance       day  duration  campaign     pdays  previous
age       1.000000  0.097783 -0.009120 -0.004648  0.004760 -0.023758  0.001288
balance   0.097783  1.000000  0.004503  0.021560 -0.014578  0.003435  0.016674
day      -0.009120  0.004503  1.000000 -0.030206  0.162490 -0.093044 -0.051710
duration -0.004648  0.021560 -0.030206  1.000000 -0.084570 -0.001565  0.001203
campaign  0.004760 -0.014578  0.162490 -0.084570  1.000000 -0.088628 -0.032855
pdays    -0.023758  0.003435 -0.093044 -0.001565 -0.088628  1.000000  0.454820
previous  0.001288  0.016674 -0.051710  0.001203 -0.032855  0.454820  1.000000

1. age - balance = 0.0977
2. day - compaign = 0.1624
3. day - pdays = -0.093044
4. pdays - previous = 0.454820 (highest correlation)

In [20]:
# Target Encoding
df.y = (df.y == 'yes').astype(int)

In [21]:
df.y.value_counts()

y
0    39922
1     5289
Name: count, dtype: int64

In [22]:
# Split the data
from sklearn.model_selection import train_test_split

In [23]:
df_full_train, df_test = train_test_split(df, test_size=0.2, random_state=42)

In [24]:
df_train, df_val = train_test_split(df_full_train, test_size=0.2, random_state=42)

In [25]:
len(df), len(df_train), len(df_val), len(df_test)

(45211, 28934, 7234, 9043)

In [26]:
df_train = df_train.reset_index(drop=True)
df_val = df_val.reset_index(drop=True)
df_test = df_test.reset_index(drop=True)

In [27]:
y_train = df_train.y.values
y_val = df_val.y.values
y_test = df_test.y.values

In [28]:
df_train.columns

Index(['age', 'job', 'marital', 'education', 'balance', 'housing', 'contact',
       'day', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'y'],
      dtype='object')

In [29]:
del df_train['y']
del df_val['y']
del df_test['y']

In [30]:
global_churn_rate = df_full_train.y.mean()
global_churn_rate

0.11606945366069453

In [31]:
df_full_train.dtypes

age           int64
job          object
marital      object
education    object
balance       int64
housing      object
contact      object
day           int64
month        object
duration      int64
campaign      int64
pdays         int64
previous      int64
poutcome     object
y             int64
dtype: object

In [32]:
df_full_train.job.value_counts()

job
blue-collar      7806
management       7500
technician       6096
admin.           4137
services         3350
retired          1815
self-employed    1286
entrepreneur     1178
unemployed       1023
housemaid        1002
student           742
unknown           233
Name: count, dtype: int64

In [33]:
df_full_train[df_full_train.job == 'blue-collar'].y.mean()

0.0722521137586472

In [34]:
df_full_train[df_full_train.job == 'student'].y.mean()

0.27897574123989216

In [35]:
df_full_train[df_full_train.job == 'entrepreneur'].y.mean()

0.0865874363327674

In [36]:
df_full_train[df_full_train.job == 'housemaid'].y.mean()

0.09381237524950099

In [37]:
numerical = list(numerical)
numerical

['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']

In [38]:
categorical = list(categorical)[:-1]
categorical

['job', 'marital', 'education', 'housing', 'contact', 'month', 'poutcome']

In [39]:
df_full_train[categorical].nunique()

job          12
marital       3
education     4
housing       2
contact       3
month        12
poutcome      4
dtype: int64

In [40]:
from IPython.display import display

In [41]:
for c in categorical:
    df_group = df_full_train.groupby(c).y.agg(['mean', 'count'])
    df_group['diff'] = df_group['mean'] - global_churn_rate
    df_group['risk'] = df_group['mean'] / global_churn_rate
    display(df_group)

mean  count      diff      risk
job                                               
admin.         0.119894   4137  0.003824  1.032947
blue-collar    0.072252   7806 -0.043817  0.622490
entrepreneur   0.086587   1178 -0.029482  0.745997
housemaid      0.093812   1002 -0.022257  0.808243
management     0.135467   7500  0.019397  1.167117
retired        0.222590   1815  0.106520  1.917727
self-employed  0.122862   1286  0.006792  1.058518
services       0.090746   3350 -0.025323  0.781827
student        0.278976    742  0.162906  2.403524
technician     0.108760   6096 -0.007310  0.937024
unemployed     0.160313   1023  0.044243  1.381180
unknown        0.111588    233 -0.004481  0.961390

mean  count      diff      risk
marital                                      
divorced  0.119720   4143  0.003651  1.031451
married   0.100578  21804 -0.015492  0.866532
single    0.147637  10221  0.031568  1.271973

mean  count      diff      risk
education                                     
primary    0.086377   5476 -0.029693  0.744183
secondary  0.105303  18670 -0.010767  0.907238
tertiary   0.148970  10539  0.032901  1.283460
unknown    0.127444   1483  0.011375  1.098001

mean  count      diff      risk
housing                                     
no       0.166407  16033  0.050337  1.433683
yes      0.075987  20135 -0.040082  0.654669

mean  count      diff      risk
contact                                       
cellular   0.148270  23464  0.032200  1.277422
telephone  0.137128   2319  0.021059  1.181431
unknown    0.038613  10385 -0.077456  0.332675

mean  count      diff      risk
month                                     
apr    0.202645   2344  0.086576  1.745895
aug    0.110954   4966 -0.005115  0.955932
dec    0.464481    183  0.348411  4.001749
feb    0.156723   2112  0.040654  1.350256
jan    0.098678   1135 -0.017391  0.850167
jul    0.091927   5537 -0.024142  0.792000
jun    0.098888   4318 -0.017181  0.851976
mar    0.516043    374  0.399973  4.445983
may    0.065219  10963 -0.050850  0.561900
nov    0.101481   3173 -0.014588  0.874315
oct    0.437396    599  0.321326  3.768396
sep    0.465517    464  0.349448  4.010678

mean  count      diff      risk
poutcome                                     
failure   0.126891   3901  0.010821  1.093229
other     0.170341   1497  0.054271  1.467575
success   0.641322   1210  0.525253  5.525332
unknown   0.090392  29560 -0.025677  0.778779

In [42]:
from sklearn.metrics import mutual_info_score

In [47]:
def mutual_info_churn_score(series):
    return mutual_info_score(series, y_train)

In [48]:
mi = df_train[categorical].apply(mutual_info_churn_score)
mi.sort_values(ascending=False).round(2)

poutcome     0.03
month        0.02
contact      0.01
housing      0.01
job          0.01
education    0.00
marital      0.00
dtype: float64

In [142]:
# 3. Variable with the biggest mutual information score is poutcome

In [49]:
# one hot encoding
from sklearn.feature_extraction import DictVectorizer

In [50]:
dv = DictVectorizer(sparse=False)

In [51]:
train_dicts = df_train[categorical + numerical].to_dict(orient='records')

In [52]:
train_dicts[0]

{'job': 'management',
 'marital': 'married',
 'education': 'tertiary',
 'housing': 'yes',
 'contact': 'cellular',
 'month': 'nov',
 'poutcome': 'unknown',
 'age': 36,
 'balance': -485,
 'day': 19,
 'duration': 278,
 'campaign': 1,
 'pdays': -1,
 'previous': 0}

In [53]:
X_train = dv.fit_transform(train_dicts)

In [54]:
val_dicts = df_val[categorical + numerical].to_dict(orient='records')

In [55]:
X_val = dv.transform(val_dicts)

In [56]:
from sklearn.linear_model import LogisticRegression

In [57]:
model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42, solver='liblinear')

In [58]:
model.intercept_[0]

-0.9569770481476176

In [59]:
len(model.coef_[0])

47

In [60]:
len(dv.get_feature_names_out())

47

In [61]:
model.coef_[0]

array([-1.11045404e-03,  1.00564988e-05, -8.42926647e-02,  2.69344331e-01,
        1.18895980e-01, -1.34521736e+00,  6.94752297e-03,  4.13800361e-03,
       -4.65326676e-01, -2.49571876e-01, -3.95694763e-02, -2.02509020e-01,
       -1.18134958e-01, -8.38842090e-01,  8.52320980e-02, -2.37508610e-01,
       -2.89497465e-01, -3.40636894e-01, -8.72487974e-02,  2.95633416e-01,
       -2.12016725e-01, -1.48421355e-01,  3.04990862e-01, -1.56973283e-01,
       -2.85590569e-02, -1.41971237e-01, -3.05622351e-01, -4.61330336e-01,
       -1.90024361e-01,  1.49816423e-02, -7.20886259e-01,  3.74559531e-01,
       -3.08456286e-01, -1.11779949e+00, -9.84200262e-01,  3.11685247e-01,
        1.36334526e+00, -5.29369995e-01, -9.60288934e-01,  8.47145453e-01,
        7.52307047e-01, -5.54022502e-04, -7.91280988e-01, -6.53256464e-01,
        1.49678700e+00, -1.00922659e+00,  7.34208065e-03])

In [62]:
model.predict(X_train)

array([0, 0, 0, ..., 0, 1, 0])

In [63]:
model.predict_proba(X_train)

array([[0.94932716, 0.05067284],
       [0.98593168, 0.01406832],
       [0.92428448, 0.07571552],
       ...,
       [0.99149133, 0.00850867],
       [0.23891672, 0.76108328],
       [0.86182184, 0.13817816]])

In [64]:
y_pred = model.predict_proba(X_val)[:,1]

In [65]:
churn_decision = (y_pred >= 0.5)
churn_decision

array([False, False, False, ...,  True, False,  True])

In [66]:
orig_accuracy = (y_val == churn_decision).mean()
print (round(orig_accuracy, 2))
# 4. Accuracy on the validation dataset is 0.9

0.9


In [67]:
accuracy_results = {}
for feature in (categorical + numerical):
    useful_features = categorical + numerical
    useful_features.remove(feature)
    train_dicts = df_train[useful_features].to_dict(orient='records')
    val_dicts = df_val[useful_features].to_dict(orient='records')
    dv = DictVectorizer(sparse=False)
    X_train = dv.fit_transform(train_dicts)
    X_val = dv.transform(val_dicts)
    model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict_proba(X_val)[:,1]
    churn_decision = (y_pred >= 0.5)
    accuracy_results[feature] = (y_val == churn_decision).mean()
accuracy_results
    

{'job': 0.9011611833010782,
 'marital': 0.902267072159248,
 'education': 0.9011611833010782,
 'housing': 0.9025435443737905,
 'contact': 0.9025435443737905,
 'month': 0.899778822228366,
 'poutcome': 0.8939729057229748,
 'age': 0.9008847110865358,
 'balance': 0.9011611833010782,
 'day': 0.9017141277301631,
 'duration': 0.8888581697539397,
 'campaign': 0.9018523638374344,
 'pdays': 0.9021288360519768,
 'previous': 0.9012994194083495}

In [68]:
df_feature_accuracy = pd.DataFrame.from_dict(accuracy_results, orient='index', columns=['accuracy'])
df_feature_accuracy['diff'] = abs(orig_accuracy - df_feature_accuracy['accuracy'])
df_feature_accuracy = df_feature_accuracy.sort_values(by='diff')
df_feature_accuracy
# 5. difference between the original accuracy and the accuracy without the feature is minimum for balance feature.

accuracy      diff
job        0.901161  0.000138
education  0.901161  0.000138
balance    0.901161  0.000138
age        0.900885  0.000138
previous   0.901299  0.000276
day        0.901714  0.000691
campaign   0.901852  0.000829
pdays      0.902129  0.001106
marital    0.902267  0.001244
month      0.899779  0.001244
housing    0.902544  0.001521
contact    0.902544  0.001521
poutcome   0.893973  0.007050
duration   0.888858  0.012165

In [69]:
# 6. 
C_accuracy_results = {}
for C in [0.01, 0.1, 1, 10, 100]:
    train_dicts = df_train[categorical + numerical].to_dict(orient='records')
    val_dicts = df_val[categorical + numerical].to_dict(orient='records')
    dv = DictVectorizer(sparse=False)
    X_train = dv.fit_transform(train_dicts)
    X_val = dv.transform(val_dicts)
    model = LogisticRegression(solver='liblinear', C=C, max_iter=1000, random_state=42)
    model.fit(X_train, y_train)
    y_pred = model.predict_proba(X_val)[:,1]
    churn_decision = (y_pred >= 0.5)
    C_accuracy_results[C] = (y_val == churn_decision).mean().round(3)

In [70]:
C_accuracy_results
# 0.1 value for C has best accuracy of 0.902

{0.01: 0.899, 0.1: 0.902, 1: 0.901, 10: 0.901, 100: 0.901}